libxrdp/xrdp_sec.c

```diff
@@ -1607,7 +1609,9 @@ xrdp_sec_process_mcs_data_CS_CORE(struct xrdp_sec* self, struct stream* s)
     in_uint8s(s, 2); /* SASSequence */
     in_uint8s(s, 4); /* keyboardLayout */
     in_uint8s(s, 4); /* clientBuild */
-    in_uint8s(s, 32); /* clientName */
+    unicode_in(s, CS_CLIENT_NAME_UNICODE_BYTES - sizeof(twchar), CS_CLIENT_NAME_UNICODE_BYTES,
+               clientName, CS_CLIENT_NAME_BYTES - sizeof(char)); /* clientName, represented as 512 bytes non-null terminated unicode string */
```

Comment:
I suspect the comment is wrong, at least the number 512. "Unicode string" is ambiguous, especially in the context that mention bytes. It should probably be called UTF-16, not Unicode. Subtracting `sizeof(twchar)` looks wrong to me. If we are dealing with UTF-16, we should be reserving two bytes, not four. And if the string we get from the client is not null-terminated, we should allocate a longer buffer for the termination, not discard the last character from the input.